### PR TITLE
[G-API] Enable 2D inputs for GIEBackend

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration
```
NV12 color format is missed.
Build configuration
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.4.1:20.04
build_image:Custom Win=openvino-2021.4.1
build_image:Custom Mac=openvino-2021.4.1

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@alalek This warnings come from `ngraph`
```
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/abs.hpp(26,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/acos.hpp(34,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/acosh.hpp(37,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/asin.hpp(37,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/asinh.hpp(37,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/atan.hpp(38,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/atanh.hpp(37,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/ceiling.hpp(28,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/constant.hpp(177,79): warning C4100: 'inputs': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/constant.hpp(177,50): warning C4100: 'outputs': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
   572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/embedding_segments_sum.hpp(69,65): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
```

Should I just suppress them for `gapi_infer_ie_test.cpp`

File: modules/gapi/CMakeLists.txt [hunk 1]
```diff
@@ -207,6 +207,10 @@ if(OPENCV_GAPI_INF_ENGINE)
   ocv_target_link_libraries(${the_module} PRIVATE ${INF_ENGINE_TARGET})
 endif()
 
+if (HAVE_NGRAPH)
+    ocv_target_link_libraries(${the_module} PRIVATE ngraph::ngraph)
```
Comment (TolyaTalamanov) [inline modules/gapi/CMakeLists.txt:211]: @alalek I noticed that `InferenceEngine` and `ngraph` are handled different way, do we need to make alias to `ngraph::ngraph` like `NGRAPH_TARGET`?

Comment (TolyaTalamanov) [reply modules/gapi/CMakeLists.txt:211]: Does it make sense not to use `InferenceEngine` separately from `ngraph` by default?

**Comment by TolyaTalamanov** [comment]
PR based on this: https://github.com/opencv/opencv/pull/20918/

File: modules/gapi/test/infer/gapi_infer_ie_test.cpp [hunk 2]
```diff
@@ -22,6 +22,10 @@
 #include "backends/ie/util.hpp"
 #include "backends/ie/giebackend/giewrapper.hpp"
 
+#ifdef HAVE_NGRAPH
+#include <ngraph/ngraph.hpp>
+#endif
```
Comment (alalek) [inline modules/gapi/test/infer/gapi_infer_ie_test.cpp:27]: > @alalek These warnings come from `ngraph`
> 
> ```
>    572>C:\Program Files (x86)\IntelSWTools\openvino_2021.4.689\deployment_tools\ngraph\include\ngraph/op/abs.hpp(26,57): warning C4100: 'visitor': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
> ```


Use this prolog code before ngraph include : https://github.com/opencv/opencv/blob/17234f82d025e3bbfbf611089637e5aa2038e7b8/modules/core/src/hal_replacement.hpp#L50-L59

and similar prolog code (check referenced file).

---

/cc @ilya-lavrenov FYI (public headers should not emit warnings)

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (alalek) [reply modules/gapi/CMakeLists.txt:211]: > separately

`find_package(InferenceEngine)` and `find_package(ngraph)`.

Check DNN module integration.

Comment (TolyaTalamanov) [reply modules/gapi/CMakeLists.txt:211]: I believe it's so similar to my implementation, isn't it? Except maybe this: https://github.com/opencv/opencv/blob/master/modules/dnn/CMakeLists.txt#L162 in DNN definitions set globally, in G-API we need this only for `opencv_test_gapi`, so far

File: modules/gapi/test/infer/gapi_infer_ie_test.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:37]: Added, works fine now, thanks!

Comment (ilya-lavrenov) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:37]: @alalek it's fixed in latest master (checked by https://github.com/openvinotoolkit/openvino/pull/8403 where I don't have to fix such issues)

Secondly, you can include IE headers as `SYSTEM`

Comment (alalek) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:37]: > as SYSTEM

Even on Windows and MSVC?

---

> in latest master

Upcoming release is 2021.4 LTS update. So we still forced to use/validate that.

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/CMakeLists.txt:211]: Is this question closed now?

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 3]
```diff
@@ -127,12 +127,13 @@ inline int toCV(IE::Precision prec) {
...
     const auto &sz = mat.size;
-
     // NB: For some reason RGB image is 2D image
     // (since channel component is not counted here).
-    // Note: regular 2D vectors also fall into this category
+    // In order to disnguish 2D tensor from image, have a look on model
+    // expected layout if it's 4D layout, most probably user passed mat is image.
     if (sz.dims() == 2 && hint == cv::gapi::ie::TraitAs::IMAGE)
     {
+        GAPI_Assert(sz.dims() == 2);
```
Comment (dmatveev) [inline modules/gapi/src/backends/ie/giebackend.cpp:136]: What's the reason to have this assert here? Under this `if` it is always `true`.

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 4]
```diff
@@ -822,6 +828,23 @@ static void configureInputInfo(const IE::InputInfo::Ptr& ii, const cv::GMetaArg
...
+    if (cv::util::holds_alternative<cv::GFrameDesc>(mm)) {
+        auto desc = cv::util::get<cv::GFrameDesc>(mm);
+        if (desc.fmt == cv::MediaFormat::NV12) {
+            info.setColorFormat(IE::ColorFormat::NV12);
+        }
+    }
+    const auto layout = ii->getTensorDesc().getLayout();
+    if (layout == IE::Layout::NCHW ||
+        layout == IE::Layout::NHWC) {
+        info.setResizeAlgorithm(IE::RESIZE_BILINEAR);
```
Comment (dmatveev) [inline modules/gapi/src/backends/ie/giebackend.cpp:843]: y don't we set resize in all other cases?

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 5]
```diff
@@ -921,19 +944,28 @@ struct Infer: public cv::detail::KernelTag {
...
                     }
-                    ii->getPreProcess().setResizeAlgorithm(IE::RESIZE_BILINEAR);
+
+                    // NB: Configure resize only for NCHW/NHWC layout,
+                    // since it isn't supposed to work with others.
+                    auto layout = ii->getTensorDesc().getLayout();
+                    if (layout == IE::Layout::NCHW ||
+                        layout == IE::Layout::NHWC) {
+                        ii->getPreProcess().setResizeAlgorithm(IE::RESIZE_BILINEAR);
+                    }
```
Comment (dmatveev) [inline modules/gapi/src/backends/ie/giebackend.cpp:968]: y does it duplicate what `configurePreProcInfo` is doing?

File: modules/gapi/src/backends/ie/giebackend.cpp [hunk 6]
```diff
@@ -1140,12 +1193,18 @@ struct InferList: public cv::detail::KernelTag {
...
 
         std::vector<std::vector<int>> cached_dims(ctx->uu.params.num_out);
         for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-            const IE::DataPtr& ie_out = ctx->uu.outputs.at(ctx->uu.params.output_names[i]);
-            cached_dims[i] = toCV(ie_out->getTensorDesc().getDims());
+            const auto& out_name = ctx->uu.params.output_names[i];
+            const auto& desc =
+                ctx->uu.params.kind == cv::gapi::ie::detail::ParamDesc::Kind::Load
+                    ? ctx->uu.net.getOutputsInfo().at(out_name)->getTensorDesc()
+                    : ctx->uu.this_network.GetOutputsInfo().at(out_name)->getTensorDesc();
```
Comment (dmatveev) [inline modules/gapi/src/backends/ie/giebackend.cpp:1206]: Maybe we should really locate this behavior in different classes wrapping this behavior for `Load` and `Import` cases, and don't have all these switches scattered in the whole file.

Comment (dmatveev) [reply modules/gapi/src/backends/ie/giebackend.cpp:1206]: But please don't overdesign it if you plan to, maybe we should put it as an input to our overall future IE backend refactoring

File: modules/gapi/test/infer/gapi_infer_ie_test.cpp [hunk 7]
```diff
@@ -22,6 +22,20 @@
 #include "backends/ie/util.hpp"
 #include "backends/ie/giebackend/giewrapper.hpp"
 
+#ifdef HAVE_NGRAPH
```
Comment (dmatveev) [inline modules/gapi/test/infer/gapi_infer_ie_test.cpp:25]: Should tests link with `ngraph::ngraph` too? I don't remember if it is the case for CMake.

Comment (TolyaTalamanov) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:25]: That's a good point, but `ngraph` is enabled by using `-DWITH_NGRAPH=ON` and this flag isn't mandatory to use `IE` itself.
So if user just do `cmake` as same as usual: `-DWITH_INF_ENGINE=ON` ngraph won't be found.


@ilya-lavrenov Could you help, please? Do you know whether `ngraph` will be mandatory part of `IE` in upcoming `OpenVINO 2.0` release?

File: modules/gapi/src/backends/ie/giebackend.cpp (hunk 6 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp:1206]: Yep, my plan was not `overdesign` implementation, I keep it as simple as possible, make sense to refactor it with `OpenVINO 2.0`

File: modules/gapi/src/backends/ie/giebackend.cpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp:968]: Yes, they do the same logic but through different interfaces, AFAIK new openvino will have only single way to set preprocessing

File: modules/gapi/src/backends/ie/giebackend.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp:843]: I believe `resize` is relevant only for `NCHW`/`NHWC` layouts.
For instance how would you resize `NC` layout (it cause `segfault` :) )

Perhaps it possible for `CHW`/`HWC`, but I don't know such models which have these input layouts

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/CMakeLists.txt:211]: @alalek Is it OK?

File: modules/gapi/src/backends/ie/giebackend.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/ie/giebackend.cpp]: You're right, missed it

**Review by dmatveev — APPROVED** [review]
LGTM but please close the linker question

File: modules/gapi/test/infer/gapi_infer_ie_test.cpp [hunk 8]
```diff
@@ -2858,6 +2872,66 @@ TEST(TestAgeGender, ThrowBlobAndInputPrecisionMismatch)
                      cv::compile_args(cv::gapi::networks(pp))));
 }
 
+#ifdef HAVE_NGRAPH
+
+TEST(Infer, ModelWith2DInputs)
+{
+    const std::string model_name   = "ModelWith2DInputs";
```
Comment (mpashchenkov) [inline modules/gapi/test/infer/gapi_infer_ie_test.cpp:2879]: Is it necessary to add some message about this model or add skip exception? We will use HAVE_NGRAPH=ON and will receive failed test without model. Should i hardcode this path for my own model?

Comment (TolyaTalamanov) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:2879]: This model is created on the fly, I didn't get your question

File: modules/gapi/include/opencv2/gapi/infer/ie.hpp [hunk 9]
```diff
@@ -80,7 +80,10 @@ struct ParamDesc {
     // NB: An optional config to setup RemoteContext for IE
     cv::util::any context_config;
 
-    size_t batch_size;
+    // NB: batch_size can't be equal to 1 by default, because some of models
+    // have 2D (Layout::NC) input and if the first dimmension not equal to 1
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/ie.hpp:84]: Done

File: modules/gapi/test/infer/gapi_infer_ie_test.cpp (hunk 8 above)
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_ie_test.cpp:2879]: Ok. I saw this.

**Review by mpashchenkov — APPROVED** [review]
LGTM

**Comment by TolyaTalamanov** [comment]
@alalek I believe it's ready to be merged

**Pull request merged**